File: leetcode/python/37. 解数独.py
#!/bin/bash python
"""
37. 解数独

思路：
经典回溯法，注意对于九宫格子的遍历方式
"""

class Solution(object):
    def isValid(self, board, i, j, value):
        """
        判断在board[i][j]的位置上，填入value是否合法
        """
        for k in range(9):
            if value == board[i][k]:
                return False
            if value == board[k][j]:
                return False
            if value == board[i//3*3 + k//3][j//3*3 + k%3]:
                return False
        return True
    
    def backTracking(self, board, points, index):
        """
        遍历当前待填入位置的集合points
        index表示当前正在填入第index个位置
        """
        if index >= len(points):
            return True
        for num in range(1, 10):
            i, j = points[index]
            value = str(num)
            if self.isValid(board, i, j, value):
                board[i][j] = value
                if self.backTracking(board, points, index + 1):
                    return True
                else:
                    board[i][j] = "."
        return False

    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """
        points = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == ".":
                    points.append((i, j))
        self.backTracking(board, points, 0)

    # ===== 其它优化版方法 =====
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        def dfs(board, pos, rows, cols, grids, index):
            """
            回溯法，返回是否成功
            """
            if index >= len(pos):
                return True
            # 遍历1-9，选取合适的值填充当前index位置
            i, j = pos[index]
            ind = i // 3 * 3 + j // 3
            for val in range(1, 10):
                if rows[i][val] == cols[j][val] == grids[ind][val] == False:
                    board[i][j] = str(val)
                    rows[i][val] = cols[j][val] = grids[ind][val] = True
                    if dfs(board, pos, rows, cols, grids, index + 1):
                        return True
                    rows[i][val] = cols[j][val] = grids[ind][val] = False
                    board[i][j] = "."
            return False
        # =========================
        if board == []:
            return
        # 使用数组hash，表示每行列九宫格出现数字的情况，第i行的数字j是否出现过
        pos = []
        rows = [[False] * 10 for _ in range(9)]
        cols = [[False] * 10 for _ in range(9)]
        grids = [[False] * 10 for _ in range(9)]
        # 遍历初始化数组hash，并保存要遍历的位置
        for i in range(9):
            for j in range(9):
                if board[i][j] == ".":
                    pos.append((i, j))
                else:
                    val = int(board[i][j])
                    ind = i // 3 * 3 + j // 3
                    rows[i][val] = cols[j][val] = grids[ind][val] = True
        # 正式遍历
        dfs(board, pos, rows, cols, grids, 0)
        return

File: leetcode/cpp/37. 解数独.cpp
class Solution {
public:
    bool isValid(vector<vector<char>>& board, int i, int j, char k) {
        for (int p = 0; p < 9; p++) {
            if (board[i][p] == k ||
                board[p][j] == k ||
                board[i/3*3 + p/3][j/3*3 + p%3] == k) {
                return false;
            }
        }
        return true;
    }

    bool perm(vector<vector<char>>& board, vector<vector<int>>& lst, int cnt) {
        if (cnt >= lst.size()) {
            return true;
        }
        for (int index = cnt; index < lst.size(); index++) {
            int x = lst[index][0];
            int y = lst[index][1];
            for (int k = '1'; k <= '9'; k++) {
                if (isValid(board, x, y, k)) {
                    board[x][y] = k;
                    if (perm(board, lst, cnt + 1)) {
                        return true;
                    }
                    board[x][y] = '.';
                }
            }
            if (board[x][y] == '.') {
                break;
            }
        }
        return false;
    }

    void solveSudoku(vector<vector<char>>& board) {
        if (board.empty()) {
            return;
        }

        vector<vector<int>> lst;
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    lst.push_back(vector<int>{i, j});
                }
            }
        }

        perm(board, lst, 0);
    }

	// ===== 其它优化方法 =====
	bool dfs(vector<vector<char>>& board, const vector<vector<int>>& pos, vector<vector<bool>>& rows, vector<vector<bool>>& cols, vector<vector<bool>>& grids, int index) {
        if (index >= pos.size()) {
            return true;
        }
        int i = pos[index][0], j = pos[index][1];
        int ind = i / 3 * 3 + j / 3;
        for (int val = 1; val <= 9; val++) {
            if (!rows[i][val] && !cols[j][val] && !grids[ind][val]) {
                board[i][j] = '0' + val;
                rows[i][val] = cols[j][val] = grids[ind][val] = true;
                if (dfs(board, pos, rows, cols, grids, index + 1)) {
                    return true;
                }
                rows[i][val] = cols[j][val] = grids[ind][val] = false;
                board[i][j] = '.';
            }
        }
        return false;
    }

    void solveSudoku(vector<vector<char>>& board) {
        // 特殊情况
        if (board.empty()) {
            return;
        }
        // 初始化
        vector<vector<int>> pos;
        vector<bool> tmp(10, false);
        vector<vector<bool>> rows(9, tmp), cols(9, tmp), grids(9, tmp);
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    pos.push_back(vector<int>{i, j});
                } else {
                    int val = board[i][j] - '0';
                    int ind = i / 3 * 3 + j / 3;
                    rows[i][val] = cols[j][val] = grids[ind][val] = true;
                }
            }
        }
        // 正式遍历
        dfs(board, pos, rows, cols, grids, 0);
        return;
    }
};